# EDA

#### Exploratory Data Analysis

- It is the process to know more about the data.
- It is exploring the data.
- It is the process to have a look upon the data from different angles and perspectives.
- It is the process to understand the different aspects of the data. 

- bullet point

# Steps

- Shape
- Basic information (datatypes)
- Column name
- Statistical summary
- Check for the null values
- Check for the duplicate values
- Check for the outliers
- Treat the null values
- Treat the duplicate values
- Treat the outliers
- Check for coorelation
- Analysis

# Importing the libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Loading the dataset

In [5]:
dt=pd.read_csv(r"C:\Users\Dell\OneDrive\Desktop\Data Science\DS Projects\Bangalore House Price Prediction\Bengaluru_House_Data.csv")

In [3]:
dt.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

   society total_sqft  bath  balcony   price  
0  Coomee        1056   2.0      1.0   39.07  
1  Theanmp       2600   5.0      3.0  120.00  
2      NaN       1440   2.0      3.0   62.00  
3  Soiewre       1521   3.0      1.0   95.00  
4      NaN       1200   2.0      1.0   51.00

# Shape

In [6]:
dt.shape

(13320, 9)

In [8]:
# - We have 13320 rows and 9 columns

# Columns

In [9]:
dt.columns

Index(['area_type', 'availability', 'location', 'size', 'society',
       'total_sqft', 'bath', 'balcony', 'price'],
      dtype='object')

# Basic Information

In [10]:
dt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   area_type     13320 non-null  object 
 1   availability  13320 non-null  object 
 2   location      13319 non-null  object 
 3   size          13304 non-null  object 
 4   society       7818 non-null   object 
 5   total_sqft    13320 non-null  object 
 6   bath          13247 non-null  float64
 7   balcony       12711 non-null  float64
 8   price         13320 non-null  float64
dtypes: float64(3), object(6)
memory usage: 936.7+ KB


In [11]:
# - Object columns- 6
# - Numeric columns- 3

# Statistical summary

In [19]:
dt.describe().style.background_gradient(cmap='Reds',low=0.5,high=1.0,axis=1)

# Check for the null values

In [20]:
dt.isna().sum()

area_type          0
availability       0
location           1
size              16
society         5502
total_sqft         0
bath              73
balcony          609
price              0
dtype: int64

- Location, size, society, bat, balcony columns have the null values

# Check for the duplicate values

In [22]:
dt.duplicated().sum()

529

In [24]:
dt[dt.duplicated()]

area_type   availability          location       size  \
971    Super built-up  Area  Ready To Move      Haralur Road      3 BHK   
1115   Super built-up  Area  Ready To Move      Haralur Road      2 BHK   
1143   Super built-up  Area  Ready To Move       Vittasandra      2 BHK   
1290   Super built-up  Area  Ready To Move      Haralur Road      2 BHK   
1394   Super built-up  Area  Ready To Move      Haralur Road      2 BHK   
...                     ...            ...               ...        ...   
13285  Super built-up  Area  Ready To Move      VHBCS Layout      2 BHK   
13299  Super built-up  Area         18-Dec        Whitefield      4 BHK   
13311            Plot  Area  Ready To Move  Ramamurthy Nagar  7 Bedroom   
13313  Super built-up  Area  Ready To Move       Uttarahalli      3 BHK   
13319  Super built-up  Area  Ready To Move      Doddathoguru      1 BHK   

       society   total_sqft  bath  balcony  price  
971    NRowse          1464   3.0      2.0   56.0  
1115       NaN         1027   2.0      2.0   44.0  
1143   Prlla C         1246   2.0      1.0   64.5  
1290       NaN         1194   2.0      2.0   47.0  
1394       NaN         1027   2.0      2.0   44.0  
...        ...          ...   ...      ...    ...  
13285  OlarkLa         1353   2.0      2.0  110.0  
13299  Prtates  2830 - 2882   5.0      0.0  154.5  
13311      NaN         1500   9.0      2.0  250.0  
13313  Aklia R         1345   2.0      1.0   57.0  
13319      NaN          550   1.0      1.0   17.0  

[529 rows x 9 columns]

- We have 529 rows as duplicate rows.

# Dropping the duplicates

In [26]:
dt.drop_duplicates(inplace=True)

In [27]:
dt.duplicated().sum()

0

- We have successfully removed the duplicates.

# Treating the null values

- 1- We can drop the null values
- 2- We can impute the null values (filling the null values with some other values)
- Before filling the null values we have to check whether a column is discrete or cataegorical

In [28]:
dt.isna().sum()

area_type          0
availability       0
location           1
size              16
society         5328
total_sqft         0
bath              73
balcony          605
price              0
dtype: int64

In [29]:
dt.nunique()

area_type          4
availability      81
location        1305
size              31
society         2688
total_sqft      2117
bath              19
balcony            4
price           1994
dtype: int64

In [30]:
# Object column will be treaed as categorical.
# Numerical column may be discrete or categorical

In [31]:
# If the column is categorical then we generally fill the null values with the mmode value.
# If the column is discrete then we generally fill the null values with mean or median.

# Loaction

In [33]:
dt.location.dtype

dtype('O')

In [45]:
dt.location.fillna(dt.location.mode()[0],inplace=True)

In [46]:
dt.location.isna().sum()

0

In [43]:
dt.location.mode()[0]

'Whitefield'

# Size

In [50]:
dt['size'].dtype

dtype('O')

In [51]:
# Mode

In [60]:
x=dt['size'].mode()
print(x,type(x))
print(x[0])

0    2 BHK
Name: size, dtype: object <class 'pandas.core.series.Series'>
2 BHK


In [54]:
dt['size'].fillna(dt['size'].mode()[0],inplace=True)

In [55]:
dt.isna().sum()

area_type          0
availability       0
location           0
size               0
society         5328
total_sqft         0
bath              73
balcony          605
price              0
dtype: int64

In [61]:
# size- column name
# size- attribute

In [64]:
dt.size

115119

# Society

- If any column has more than 30% values as null values we drop that column.

In [69]:
dt.society.isna().sum()*100/len(dt)

41.654288171370496

In [70]:
# Since 41% values are null values so we can drop the societ column.

In [72]:
dt.drop('society',axis=1,inplace=True)

In [73]:
dt.head()

area_type   availability                  location       size  \
0  Super built-up  Area         19-Dec  Electronic City Phase II      2 BHK   
1            Plot  Area  Ready To Move          Chikka Tirupathi  4 Bedroom   
2        Built-up  Area  Ready To Move               Uttarahalli      3 BHK   
3  Super built-up  Area  Ready To Move        Lingadheeranahalli      3 BHK   
4  Super built-up  Area  Ready To Move                  Kothanur      2 BHK   

  total_sqft  bath  balcony   price  
0       1056   2.0      1.0   39.07  
1       2600   5.0      3.0  120.00  
2       1440   2.0      3.0   62.00  
3       1521   3.0      1.0   95.00  
4       1200   2.0      1.0   51.00

In [74]:
dt.isna().sum()

area_type         0
availability      0
location          0
size              0
total_sqft        0
bath             73
balcony         605
price             0
dtype: int64

# bath

In [75]:
dt.bath.dtype

dtype('float64')

In [76]:
dt.nunique()

area_type          4
availability      81
location        1305
size              31
total_sqft      2117
bath              19
balcony            4
price           1994
dtype: int64

In [78]:
dt.bath.fillna(dt.bath.mode()[0],inplace=True)

In [79]:
dt.isna().sum()

area_type         0
availability      0
location          0
size              0
total_sqft        0
bath              0
balcony         605
price             0
dtype: int64

# balcony

In [81]:
dt.balcony.dtype

dtype('float64')

In [82]:
dt.balcony.fillna(dt.balcony.mode()[0],inplace=True)

In [83]:
dt.isna().sum()

area_type       0
availability    0
location        0
size            0
total_sqft      0
bath            0
balcony         0
price           0
dtype: int64

- We have successfully treated the null values.

In [85]:
# data- algorithm

In [88]:
# x=dt.location.mode()
# x.iloc[0]